File: app/src/main/java/bassamalim/hidaya/activities/AthkarListActivity.java
package bassamalim.hidaya.activities;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.SearchView;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.room.Room;

import java.util.ArrayList;
import java.util.List;

import bassamalim.hidaya.R;
import bassamalim.hidaya.adapters.AthkarListAdapter;
import bassamalim.hidaya.database.AppDatabase;
import bassamalim.hidaya.database.dbs.AthkarDB;
import bassamalim.hidaya.database.dbs.ThikrsDB;
import bassamalim.hidaya.databinding.ActivityAthkarListBinding;
import bassamalim.hidaya.models.AthkarItem;
import bassamalim.hidaya.other.Global;
import bassamalim.hidaya.other.Utils;

public class AthkarListActivity extends AppCompatActivity {

    private ActivityAthkarListBinding binding;
    private RecyclerView recyclerView;
    private AthkarListAdapter adapter;
    private int category;
    private String action;
    private AppDatabase db;
    private String language;
    private List<String> names;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Utils.onActivityCreateSetTheme(this);
        language = Utils.onActivityCreateSetLocale(this);
        binding = ActivityAthkarListBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        binding.home.setOnClickListener(v -> onBackPressed());

        db = Room.databaseBuilder(getApplicationContext(), AppDatabase.class,
                "HidayaDB").createFromAsset("databases/HidayaDB.db").allowMainThreadQueries()
                .build();

        action = getIntent().getAction();
        if (action.equals("all"))
            binding.topBarTitle.setText(getString(R.string.all_athkar));
        else if (action.equals("favorite"))
            binding.topBarTitle.setText(getString(R.string.favorite_athkar));
        else {
            category = getIntent().getIntExtra("category", 0);

            String topBarTitle = language.equals("en") ? db.athkarCategoryDao()
                    .getNameEn(category) : db.athkarCategoryDao().getName(category);
            binding.topBarTitle.setText(topBarTitle);
        }

        setupRecycler();

        setSearchListeners();
    }

    private List<AthkarDB> getData() {
        names = language.equals("en") ? db.athkarDao().getNamesEn() : db.athkarDao().getNames();

        switch (action) {
            case "all": return db.athkarDao().getAll();
            case "favorite": return db.athkarDao().getFavorites();
            default: return db.athkarDao().getList(category);
        }
    }

    private List<AthkarItem> makeButtons(List<AthkarDB> athkar) {
        Log.d(Global.TAG, "SIZE: " + athkar.size());
        List<AthkarItem> buttons = new ArrayList<>();

        List<Integer> favs = db.athkarDao().getFavs();

        for (int i = 0; i < athkar.size(); i++) {
            AthkarDB thikr = athkar.get(i);

            if (language.equals("en") && !hasEn(thikr))
                continue;

            View.OnClickListener clickListener = v -> {
                Intent intent = new Intent(this, AthkarViewer.class);
                intent.setAction(action);
                intent.putExtra("thikr_id", thikr.getAthkar_id());
                startActivity(intent);
            };

            int fav = action.equals("favorite") ? 1 : favs.get(thikr.getAthkar_id());

            buttons.add(new AthkarItem(thikr.getAthkar_id(), thikr.getCategory_id(),
                    names.get(thikr.getAthkar_id()), fav, clickListener));
        }
        return buttons;
    }

    private boolean hasEn(AthkarDB thikr) {
        List<ThikrsDB> ts = db.thikrsDao().getThikrs(thikr.getAthkar_id());
        for (int i = 0; i < ts.size(); i++) {
            ThikrsDB t = ts.get(i);
            if (t.getText_en() != null && t.getText_en().length() > 1)
                return true;
        }
        return false;
    }

    private void setupRecycler() {
        recyclerView = findViewById(R.id.recycler);
        LinearLayoutManager layoutManager = new LinearLayoutManager(this);
        recyclerView.setLayoutManager(layoutManager);
        adapter = new AthkarListAdapter(this, makeButtons(getData()));
        recyclerView.setAdapter(adapter);
    }

    private void setSearchListeners() {
        binding.searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                adapter.filter(query);
                return true;
            }
            @Override
            public boolean onQueryTextChange(String newText) {
                adapter.filter(newText);
                return true;
            }
        });
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        binding = null;
        recyclerView.setAdapter(null);
        adapter = null;
    }
}

File: app/src/main/java/bassamalim/hidaya/activities/AthkarListActivity.kt
package bassamalim.hidaya.activities

import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.items
import androidx.compose.runtime.Composable
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.input.TextFieldValue
import androidx.preference.PreferenceManager
import bassamalim.hidaya.R
import bassamalim.hidaya.database.AppDatabase
import bassamalim.hidaya.database.dbs.AthkarDB
import bassamalim.hidaya.models.AthkarItem
import bassamalim.hidaya.ui.components.*
import bassamalim.hidaya.ui.theme.AppTheme
import bassamalim.hidaya.utils.ActivityUtils
import bassamalim.hidaya.utils.DBUtils
import com.google.gson.Gson

class AthkarListActivity : ComponentActivity() {

    private lateinit var pref: SharedPreferences
    private val gson = Gson()
    private var category = 0
    private lateinit var action: String
    private lateinit var db: AppDatabase
    private lateinit var language: String

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        language = ActivityUtils.onActivityCreateSetLocale(this)

        db = DBUtils.getDB(this)
        pref = PreferenceManager.getDefaultSharedPreferences(this)

        action = intent.action!!

        val title: String
        when (action) {
            "all" -> title = getString(R.string.all_athkar)
            "favorite" -> title = getString(R.string.favorite_athkar)
            else -> {
                category = intent.getIntExtra("category", 0)
                title =
                    if (language == "en") db.athkarCategoryDao().getNameEn(category)
                    else db.athkarCategoryDao().getName(category)
            }
        }

        setContent {
            AppTheme {
                UI(title)
            }
        }
    }

    private val data: List<AthkarDB>
        get() {
            return when (action) {
                "all" -> db.athkarDao().getAll()
                "favorite" -> db.athkarDao().getFavorites()
                else -> db.athkarDao().getList(category)
            }
        }

    private fun getItems(athkar: List<AthkarDB>): List<AthkarItem> {
        val items: MutableList<AthkarItem> = ArrayList()

        for (i in athkar.indices) {
            val thikr = athkar[i]

            if (language == "en" && !hasEn(thikr)) continue

            val name =
                if (language == "en") thikr.name_en!!
                else thikr.name!!

            items.add(
                AthkarItem(
                    thikr.id, thikr.category_id, name, mutableStateOf(thikr.favorite)
                )
            )
        }
        return items
    }

    private fun hasEn(thikr: AthkarDB): Boolean {
        val ts = db.thikrsDao().getThikrs(thikr.id)
        for (i in ts.indices) {
            val t = ts[i]
            if (t.getTextEn() != null && t.getTextEn()!!.length > 1) return true
        }
        return false
    }

    private fun updateFavorites() {
        val favAthkar = db.athkarDao().getFavs()
        val athkarJson = gson.toJson(favAthkar)
        pref.edit()
            .putString("favorite_athkar", athkarJson)
            .apply()
    }

    @Composable
    private fun UI(title: String) {
        MyScaffold(title) {
            val textState = remember { mutableStateOf(TextFieldValue("")) }

            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(it)
            ) {
                SearchView (
                    state = textState,
                    hint = stringResource(id = R.string.athkar_hint)
                )

                MyLazyColumn(
                    lazyList = {
                        items(
                            items = getItems(data).filter { item ->
                                item.name.contains(textState.value.text, ignoreCase = true)
                            }
                        ) { item ->
                            MyBtnSurface(
                                text = item.name,
                                iconBtn = {
                                    MyFavBtn(
                                        fav = item.favorite.value
                                    ) {
                                        if (item.favorite.value == 0) {
                                            db.athkarDao().setFav(item.id, 1)
                                            item.favorite.value = 1
                                        }
                                        else if (item.favorite.value == 1) {
                                            db.athkarDao().setFav(item.id, 0)
                                            item.favorite.value = 0
                                        }
                                        updateFavorites()
                                    }
                                }
                            ) {
                                val intent = Intent(
                                    this@AthkarListActivity, AthkarViewer::class.java
                                )
                                intent.action = action
                                intent.putExtra("thikr_id", item.id)
                                startActivity(intent)
                            }
                        }
                    }
                )
            }
        }
    }

}